src/types/refs.rs

```diff
@@ -318,3 +329,63 @@ impl<'a, T: NativeRepr> ScriptRef<'a, T> {
         !self.0.ref_.is_null()
     }
 }
+
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct SharedPtr<T>(red::SharedPtrBase<T>);
+
+impl<T> Clone for SharedPtr<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        self.inc_strong();
+        unsafe { ptr::read(self) }
+    }
+}
+
+impl<T: Default + NativeRepr> SharedPtr<T> {
+    #[must_use]
+    pub fn new_with(value: T) -> Self {
+        let mut this = red::SharedPtrBase::<T>::default();
+        let refcount = RefCount::new();
+        this.refCount = refcount.0 as *mut red::RefCnt;
+        this.instance = Box::leak(Box::new(value)) as *const _ as *mut _;
```

Comment:
```suggestion
        this.instance = Box::leak(Box::new(value)) as *const _ as *mut _;
```
this is pretty bad, using this constructor repeatedely will leak lots of memory

Reply:
isn't this memory eventually reclaimed in `Drop` impl ? (`ptr::drop_in_place(ptr_instance);`)

Reply:
if not, what would be the appropriate approach for this @jac3km4 ? this PR has been sitting for almost a year i'm longing to finish it :) @psiberx @wopss if ever you have suggestions too I'm all ears !

Reply:
From what I read on the internet, this part is partially correct, only that in the `Drop` implementation `ptr::drop_in_place` will **not** destroy the heap object. 

What I saw, is that you do this:
```suggestion
        this.instance = Box::into_raw(value);
```

and then in `Drop` instead of `ptr::drop_in_place`, you do `Box::from_raw`.

See https://github.com/rust-lang/rust/blob/855e0fe46e68d94e9f6147531b75ac2d488c548e/library/alloc/src/boxed.rs#L1126-L1147.

Reply:
`SharedPtr<T>` only makes sense if this specific `SharedPtr<T>` is already used by the game (not even for every `T` known to the engine), and only if the right allocator is used for `instance` of this specific `SharedPtr<T>` (or sometimes `T`) case.

Reply:
thanks @wopss i think i understand your part and will make the appropriate changes.

following psiberx comment which i understand only partially @jac3km4, in Rust does it simply mean:
- making `SharedPtr<T>(red::SharedPtrBase<T>)` into `SharedPtr<T, A>(red::SharedPtrBase<T>, PhantomData<A>);`
  then, e.g. turning `impl<T> Clone for SharedPtr<T>` into `impl<T> Clone for SharedPtr<T, crate::types::IAllocator>` (for all the impl out there)
- or i need to take the extra leap and use RED4ext methods from `bindgen` like maybe `IAllocator_Alloc` / `IAllocator_AllocAligned` and `IAllocator_Free` (in `IAllocator__bindgen_vtable`).

these things are still out of my league, so ur help is very much appreciated ❤️

Reply:
What @psiberx pointed out is that the memory for the shared object is Allocated using Rust's default allocator, but the memory in the game engine is allocated using a custom allocator, through `IAllocator` interfaces and friends.

If you allocate memory in Rust but then pass the ownership to the engine (for example, by putting the object into a game-owned `DynArray`), the engine won't recognize where that memory came from and might panic.

So, the correct approach is: instead of using `Box::new`, use the allocator from the engine (`IAllocator_*`).

As for the `SharedPtr<T, A>` idea: I'm not sure if that's a good option, my Rust knowledge is 0. @jac3km4, might be better suited  to judge that part. In our C++ code, we get the allocator from type aliases or methods like `GetAllocator`.